x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/InternalUsers.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.user;
+
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+public class InternalUsers {
+
+    private record UserInstance(String name, User user, @Nullable RoleDescriptor role, Predicate<User> isUserPredicate) {
+        private UserInstance {
+            // Check that the parameters align
+            assert User.isInternal(user) : "User " + user + " is not internal";
+            // The user principal should match the official name
+            assert name.equals(user.principal())
+                : "User " + user + " has a principal [" + user.principal() + "] that dopes not match the provided name [" + name + "]";
+            // The role descriptor should match the user name as well
+            assert role == null || name.equals(role.getName())
+                : "Internal user " + user + " should have a role named [" + name + "] but was [" + role.getName() + "]";
+            // : The user object should match as an instance of the user.
+            assert isUserPredicate.test(user) : "User " + user + " does not match provided predicate";
+        }
+
+        public boolean is(User u) {
+            return isUserPredicate.test(u);
+        }
+    }
+
+    private static final List<UserInstance> INTERNAL_USERS = List.of(
+        new UserInstance(SystemUser.NAME, SystemUser.INSTANCE, null, SystemUser::is),
+        new UserInstance(XPackUser.NAME, XPackUser.INSTANCE, XPackUser.ROLE_DESCRIPTOR, XPackUser::is),
+        new UserInstance(XPackSecurityUser.NAME, XPackSecurityUser.INSTANCE, XPackSecurityUser.ROLE_DESCRIPTOR, XPackSecurityUser::is),
+        new UserInstance(
+            SecurityProfileUser.NAME,
+            SecurityProfileUser.INSTANCE,
+            SecurityProfileUser.ROLE_DESCRIPTOR,
+            SecurityProfileUser::is
+        ),
+        new UserInstance(AsyncSearchUser.NAME, AsyncSearchUser.INSTANCE, AsyncSearchUser.ROLE_DESCRIPTOR, AsyncSearchUser::is),
+        new UserInstance(CrossClusterAccessUser.NAME, CrossClusterAccessUser.INSTANCE, null, CrossClusterAccessUser::is)
```

Comment:
Passing `null` here is a bit hacky. Maybe we should at least comment about why this is the case.
The reason why crossClusterAccess user does not resolve roles like other internal users is because it never runs on its own. Its usage is mostly on the fulfilling cluster side where it is always bound by the cluster API key. Therefore it's role is resolved differently.

Reply:
Is it important that we never resolve the role? Not having a role descriptor for `SystemUser` and `CrossClusterAccessUser` is a hack. I think they should have roles.

SystemUser is harder to fix because it currently has a predicate instead of a role, so there's a bit of work to do to if we want to change that, but CrossClusterAccessUser does have a role - why do we insist on not returning it?

Reply:
Probably not critical. But the current design expects cross_cluster_access user to not go through this path. So I think it's better to keep this expectation for now. We could figure out something later if necessary.